### PR TITLE
Feature zip iterator and zip container view.

### PR DESCRIPTION
**Pull request by rrahn**
Main features:
   * new ZipIterator similar to the boost::zip_iterator.
   * ZipView that subclasses the ContainerView to make the zip interface available for the container based design of SeqAn, e.g. to call it with functions as forEach, iterate etc..

In general the classes use variadic templates and the std::tuple to zip different container types together.
With the variadic templates the entire code becomes very concise!!!
@esiragusa @h-2  please have a look and comment on the design and stuff that can be improved.
I added a howto demo to simply give a show case and to test if it works.
I wrote some tests but will include them first when we are sure about the design.

### CODE REVIEW

File: include/seqan/basic/iterator_zipped.h [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+{
+
+// ============================================================================
+// Forwards
+// ============================================================================
+
+// This function enables us to use the initializer-list trick to expand a parameter pack,
+// in order to call a function on each element of the parameter pack.
+template <typename T> inline void
+_seqanUnpackFunc(std::initializer_list<T> const /*unusued*/)
```
Comment (h-2) [inline include/seqan/basic/iterator_zipped.h]: inline void into extra line

File: include/seqan/basic/iterator_zipped.h [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+{
+    typedef std::tuple<typename Reference<TIteratorPack>::Type...>  Type;
+};
+
+// Unclear why we need that! It should be possible to let the generic implementation
+// in iterator_base.h to take care of it. Apparently, it is the same semantic.
+template <typename... TIteratorPack>
+struct Reference<Iter<std::tuple<TIteratorPack...>, ZipIterator> const> :
+    Reference<Iter<std::tuple<TIteratorPack...>, ZipIterator> > {};
+
```
Comment (h-2) [inline include/seqan/basic/iterator_zipped.h]: shouldnt it be const_reference then as well?

File: include/seqan/sequence/container_view_zipped.h [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    template <typename... TOtherContPack>
+    ContainerView(TOtherContPack && ...contArgs)
+    {
+        static_assert(sizeof...(TOtherContPack) > 0, "Requires at least one argument!");
+        static_assert(sizeof...(TOtherContPack) == sizeof...(TContPack), "Number of parameters don't match!");
+
+        // TODO: We need to assert if lenght of containers differ!
+
```
Comment (h-2) [inline include/seqan/sequence/container_view_zipped.h]: doesnt the above do that?

File: include/seqan/sequence/container_view_zipped.h [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        _end(end)
+    {}
+
+    // ------------------------------------------------------------------------
+    // Operator =
+    // ------------------------------------------------------------------------
+
+    template <typename TOtherContainer>
+    SEQAN_HOST_DEVICE inline
+    ContainerView &
```
Comment (h-2) [inline include/seqan/sequence/container_view_zipped.h]: SEQAN_HOST_DEVICE should go away IIANM.

File: include/seqan/sequence/container_view_zipped.h [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+};
+
+template <typename... TContPack, typename TSpec>
+struct Reference<ContainerView<std::tuple<TContPack...>, ZipContainer<TSpec> > const>
+{
+    typedef typename Iterator<ContainerView<std::tuple<TContPack...>, ZipContainer<TSpec> > const, Standard>::Type TIter_;
+    typedef typename Reference<TIter_>::Type                                                                       Type;
+};
+
+
```
Comment (h-2) [inline include/seqan/sequence/container_view_zipped.h]: double newline

File: include/seqan/sequence/container_view_zipped.h (hunk 4 above)
Comment (rrahn) [inline include/seqan/sequence/container_view_zipped.h]: yes, for sure.

File: include/seqan/sequence/container_view_zipped.h (hunk 3 above)
Comment (rrahn) [inline include/seqan/sequence/container_view_zipped.h]: No what I meant is that the length of all the containers that are zipped should have the same length (the same number of elements). The statement above just checks that the number of containers is the same.

File: include/seqan/basic/iterator_zipped.h (hunk 2 above)
Comment (rrahn) [inline include/seqan/basic/iterator_zipped.h]: Nop. The reference is the reference to the value of the underlying container an iterator is pointing to. Hence the const qualification depends on the constness of the container not the iterator itself. 
However, there is a generic MF that should do the above thing but it does not work for whatever reasons.

**Comment by rrahn** [comment]
Ok I finalised it. @h-2 if everything is ok, can you merge it?

**Comment by rrahn** [comment]
Oh shit I messed up something here.

**Comment by rrahn** [comment]
Please hold it.

**Comment by rrahn** [comment]
In favour of #1388

**Pull request closed without merging**